File: problems12/lab12_b.cpp
#include <vector>
#include <iostream>
int main() {
    long long n;
    std::cin >> n;
    std::vector<long long> seq, lis;
    for (long long i = 0; i < n; ++i) {
        long long temp;
        std::cin >> temp;
        seq.push_back(temp);
    }
    std::vector<int> parent(seq.size());
    lis.push_back(0);
    long long aim, index;
    for (long long i = 1; i < seq.size(); i++) {
        if (seq[lis.back()] < seq[i]) {
            parent[i] = lis.back();
            lis.push_back(i);
            continue;
        }
        for (aim = 0, index = lis.size() - 1; aim < index;) {
            int c = (aim + index) / 2;
            if (seq[lis[c]] < seq[i]) {aim = c + 1;}
            else {index = c;}
        }
        if (seq[i] < seq[lis[aim]]) {
            if (aim > 0) {parent[i] = lis[aim - 1];}
            lis[aim] = i;
        }
    }
    for (aim = lis.size(), index = lis.back(); aim--; index = parent[index]) {lis[aim] = index;}
    std::cout << lis.size() << std::endl;
    for (long long li : lis) {std::cout << seq[li] << " ";}
    return 0;
}

File: problems12/lab12_b.py
from sys import setrecursionlimit
import threading
setrecursionlimit(10 ** 9)
threading.stack_size(3 * 67108864)

def main(): #Knuth LIS-binsearch algo O(nlogn)
    def binsearch():
        left, right = 1, largest + 1
        while right - left > 1:
            mid = (left + right) // 2
            if data[sub[mid]] < data[i]: left = mid
            else: right = mid
        return left
    n = int(input())
    data = [None] + list(map(int, input().split()))
    # [list] sub[i]: index(min last element) in sequence of len == i
    # [list] parent[k]: index of predecessor of sub[k] in LIS ending at sub[k]
    sub, parent = [None] * (n + 1), [None] * (n + 1)
    largest = 0
    for i in range(1, n + 1):
        aim = 0 if (largest == 0 or data[sub[1]] >= data[i]) else binsearch()
        parent[i] = sub[aim]
        if aim == largest or data[sub[aim + 1]] > data[i]:
            sub[aim + 1] = i
            largest = max(largest, aim + 1)
    ans, index = [], sub[largest]
    while largest > 0:
        ans.append(data[index])
        index = parent[index]
        largest -= 1
    print(len(ans))
    print(*ans[::-1])

thread = threading.Thread(target=main)
thread.start()